File: M/85-maximal-rectangle/main.go
package main

import "fmt"

func maximalRectangle(matrix [][]byte) int {
	buf := make([][][]int, len(matrix))
	for i := 0; i < len(matrix); i++ {
		buf[i] = make([][]int, len(matrix[0]))
	}

	max := 0
	for i := len(matrix) - 1; i >= 0; i-- {
		for j := len(matrix[0]) - 1; j >= 0; j-- {
			if matrix[i][j] == '0' {
				buf[i][j] = []int{0, 0}
				continue
			}
			r, c := 0, 0
			if i < len(matrix)-1 {
				c = buf[i+1][j][1]
			}
			if j < len(matrix[0])-1 {
				r = buf[i][j+1][0]
			}
			buf[i][j] = []int{r + 1, c + 1}

			rmin := len(matrix[0])
			for n := 0; n < c+1; n++ {
				if rmin > buf[i+n][j][0] {
					rmin = buf[i+n][j][0]
				}
				if max < rmin*(n+1) {
					max = rmin * (n + 1)
				}
			}
		}
	}
	return max
}

func main() {
	fmt.Println()
}

File: M/85-maximal-rectangle/main.rs
impl Solution {
  pub fn maximal_rectangle(matrix: Vec<Vec<char>>) -> i32 {
    let mut height: Vec<i32> = vec![0; matrix[0].len() + 1];
    let mut ans: i32 = 0;
    for r in 0..matrix.len() {
      let mut stack: Vec<(i32, i32)> = vec![(-1, -1)];
      for c in 0..matrix[0].len() {
        if matrix[r][c] == '0' {
          height[c] = 0;
        } else {
          height[c] += 1;
        }
      }
      for c in 0..height.len() {
        while stack.len() > 1 && stack[stack.len() - 1].0 >= height[c] {
          let (h, i) = stack.pop().unwrap();
          ans = ans.max(h * (c as i32 - stack[stack.len() - 1].1 - 1));
        }
        stack.push((height[c], c as i32));
      }
    }
    ans
  }
}
